README.md

# [MOVED] Analytics JS integration mixpanel

**This repository has been moved to the open source [Analytics JS Integrations](https://github.com/segmentio/analytics.js-integrations).**

If you want to fork, contribute, or open issues, please do it in the new repository. Existing issues/pull requests will be addressed in the new location.

* [New location](https://github.com/segmentio/analytics.js-integrations/tree/master/integrations/mixpanel)
* Last release for this repository: [3.0.1](https://github.com/segment-integrations/analytics.js-integration-mixpanel/releases/tag/3.0.1)
* [Commit](https://github.com/segmentio/analytics.js-integrations/commit/9753aca8896f90f90f7d4b2d1a578a2a33bc1891)
* [Previous version of this README](README-OLD.md)

Released under the [MIT license](LICENSE).


HISTORY.md

3.0.0 / 2017-10-19
==================

  * Parity rewrite (#29)
  * Bump a.js-int-tester version to ^3.1.0 (#28)
  * Pin karma, karma-mocha dev dependencies (#27)

2.2.0 / 2016-08-16
==================

  * update snippet

2.1.0 / 2016-08-12
==================

  * tag mp_lib with Segment: web

2.0.0 / 2016-06-21
==================

  * Remove Duo compatibility
  * Add CI setup (coverage, linting, cross-browser compatibility, etc.)
  * Update eslint configuration


1.5.0 / 2016-05-12
==================

  * Update Page Call Logic with consolidatedPageCalls setting

1.4.2 / 2016-05-07
==================

  * Bump Analytics.js core, tester, integration to use Facade 2.x

1.4.1 / 2016-02-03
==================

  * fix multiple page call issue

1.4.0 / 2015-12-09
==================

  * add persistence config

1.3.2 / 2015-11-03
==================

  * make mixpanel es5 friendly
  * Dont set peopleProperties in track calls.

1.3.1 / 2015-10-29
==================

  * Dont set peopleProperties in track calls. Dont map special traits in track, people should use identify instead.

1.3.0 / 2015-10-27
==================



1.3.0 / 2015-10-27
==================

  * refactor super people props

1.2.1 / 2015-09-08
==================

  * adding check for no Mixpanel.superProperties object closes #11

1.2.0 / 2015-09-08
==================

  * Allow explicit declaration of people/super properties ([More Info](https://github.com/segment-integrations/analytics.js-integration-mixpanel/pull/10))
  * stub every possible mixpanel method

1.1.0 / 2015-08-25
==================

  * Merge pull request #6 from segment-integrations/people-super-properties
  * added enhancedTrack support

1.0.5 / 2015-08-05
==================

  * Send arrays through to Mixpanel

1.0.4 / 2015-06-30
==================

  * Replace analytics.js dependency with analytics.js-core

1.0.3 / 2015-06-24
==================

  * Bump analytics.js-integration version

1.0.2 / 2015-06-24
==================

  * Bump analytics.js-integration version

1.0.1 / 2015-06-10
==================

  * Bump dependency versions

1.0.0 / 2015-06-09
==================

  * Initial commit :sparkles:


test/index.test.js
'use strict';

var Analytics = require('@segment/analytics.js-core').constructor;
var integration = require('@segment/analytics.js-integration');
var iso = require('@segment/to-iso-string');
var sandbox = require('@segment/clear-env');
var tester = require('@segment/analytics.js-integration-tester');
var Mixpanel = require('../lib/');

describe('Mixpanel', function() {
  var analytics;
  var mixpanel;
  var options = {
    token: 'x',
    cookieName: 'y',
    crossSubdomainCookie: true,
    secureCookie: true,
    consolidatedPageCalls: false,
    trackCategorizedPages: true,
    trackNamedPages: true
  };

  beforeEach(function() {
    analytics = new Analytics();
    mixpanel = new Mixpanel(options);
    analytics.use(Mixpanel);
    analytics.use(tester);
    analytics.add(mixpanel);
  });

  afterEach(function() {
    analytics.restore();
    analytics.reset();
    mixpanel.reset();
    sandbox();
  });

  it('should have the right settings', function() {
    analytics.compare(Mixpanel, integration('Mixpanel')
      .global('mixpanel')
      .option('token', '')
      .option('crossSubdomainCookie', false)
      .option('secureCookie', false)
      .option('nameTag', true)
      .option('pageview', false)
      .option('people', false)
      .option('token', '')
      .option('trackAllPages', false)
      .option('persistence', 'cookie')
      .option('trackNamedPages', false)
      .option('consolidatedPageCalls', true)
      .option('setAllTraitsByDefault', true)
      .option('trackCategorizedPages', false));
  });

  describe('before loading', function() {
    beforeEach(function() {
      analytics.stub(mixpanel, 'load');
    });

    describe('#initialize', function() {
      it('should create window.mixpanel', function() {
        analytics.assert(!window.mixpanel);
        analytics.initialize();
        analytics.page();
        analytics.assert(window.mixpanel);
      });

      it('should call #load', function() {
        analytics.initialize();
        analytics.page();
        analytics.called(mixpanel.load);
      });

      it('should lowercase event increments', function() {
        mixpanel.options.eventIncrements = ['A', 'b', 'c_'];
        analytics.initialize();
        analytics.page();
        analytics.deepEqual(mixpanel.options.eventIncrements, ['a', 'b', 'c_']);
      });
    });
  });

  describe('loading', function() {
    it('should load', function(done) {
      analytics.load(mixpanel, done);
    });
  });

  describe('after loading', function() {
    beforeEach(function(done) {
      analytics.once('ready', done);
      analytics.initialize();
      analytics.page();
    });

    describe('config', function() {
      it('should tag mp_lib as Segment: web', function() {
        analytics.initialize();
        analytics.page();
        analytics.assert(window.mixpanel.config.loaded.toString().slice(98, -2), 'mixpanel.register({"mp_lib":"Segment: web"})');
      });
    });

    describe('#page', function() {
      beforeEach(function() {
        analytics.stub(window.mixpanel, 'track');
        analytics.stub(window.mixpanel.people, 'set');
      });

      it('should not track anonymous pages by default', function() {
        analytics.page();
        analytics.didNotCall(window.mixpanel.track);
      });

      it('should track anonymous pages when the option is on', function() {
        mixpanel.options.trackAllPages = true;
        analytics.page();
        analytics.called(window.mixpanel.track, 'Loaded a Page');
      });

      it('should track named pages by default', function() {
        analytics.page('Name');
        analytics.called(window.mixpanel.track, 'Viewed Name Page');
      });

      it('should track named pages with categories', function() {
        analytics.page('Category', 'Name');
        analytics.called(window.mixpanel.track, 'Viewed Category Name Page');
      });

      it('should track categorized pages by default', function() {
        analytics.page('Category', 'Name');
        // This test was incorrectly testing this before, as it was passing based on the multiple call sending bug. The correct new call would be the following as noted in our docs:
        analytics.called(window.mixpanel.track, 'Viewed Category Name Page');
        analytics.didNotCall(window.mixpanel.track, 'Viewed Category Page');
        analytics.didNotCall(window.mixpanel.track, 'Viewed Name Page');
      });

      it('should not track category pages when the option is off', function() {
        mixpanel.options.trackNamedPages = false;
        mixpanel.options.trackCategorizedPages = false;
        analytics.page('Name');
        analytics.page('Category', 'Name');
        analytics.didNotCall(window.mixpanel.track);
      });

      // Exercise a bug where we set people properties in track calls. https://segment.phacility.com/T903
      it('will not track page props as a people property', function() {
        mixpanel.options.people = true;
        analytics.page('Name');
        analytics.called(window.mixpanel.track, 'Viewed Name Page');
        analytics.didNotCall(window.mixpanel.people.set, { $name: 'Name' });
      });

      it('should not send multiple page calls', function() {
        mixpanel.options.trackAllPages = true;
        mixpanel.options.trackNamedPages = true;
        analytics.page('Teemo');
        analytics.called(window.mixpanel.track, 'Loaded a Page');
        analytics.didNotCall(window.mixpanel.track, 'Viewed Teemo Page');
      });

      // consolidatedPageCalls = true
      it('should not send multiple page calls', function() {
        mixpanel.options.consolidatedPageCalls = true;
        analytics.page('Teemo');
        analytics.called(window.mixpanel.track, 'Loaded a Page', {
          name: 'Teemo',
          path: '/context.html',
          referrer: document.referrer,
          search: '',
          title: document.title,
          url: window.location.href
        });
        analytics.didNotCall(window.mixpanel.track, 'Viewed Teemo Page');
      });
    });

    describe('#identify', function() {
      beforeEach(function() {
        analytics.stub(window.mixpanel, 'identify');
        analytics.stub(window.mixpanel, 'register');
        analytics.stub(window.mixpanel, 'name_tag');
        analytics.stub(window.mixpanel.people, 'set');
        analytics.stub(window.mixpanel.people, 'union');
      });

      it('should send an id', function() {
        analytics.identify('id');
        analytics.called(window.mixpanel.identify, 'id');
        analytics.called(window.mixpanel.register, { id: 'id' });
      });

      it('should send traits', function() {
        analytics.identify({ trait: true });
        analytics.called(window.mixpanel.register, { trait: true });
      });

      it('should send an id and traits', function() {
        analytics.identify('id', { trait: true });
        analytics.called(window.mixpanel.identify, 'id');
        analytics.called(window.mixpanel.register, { trait: true, id: 'id' });
      });

      it('should use an id as a name tag', function() {
        analytics.identify('id');
        analytics.called(window.mixpanel.name_tag, 'id');
      });

      it('should prefer a username as a name tag', function() {
        analytics.identify('id', { username: 'username' });
        analytics.called(window.mixpanel.name_tag, 'username');
      });

      it('should prefer an email as a name tag', function() {
        analytics.identify('id', {
          username: 'username',
          email: 'name@example.com'
        });
        analytics.called(window.mixpanel.name_tag, 'name@example.com');
      });

      it('should send traits to Mixpanel People', function() {
        mixpanel.options.people = true;
        analytics.identify({ trait: true });
        analytics.called(window.mixpanel.people.set, { trait: true });
      });

      it('should set super properties from the mixpanel.options.superProperties object if setAllTraitsByDefault is false', function() {
        mixpanel.options.setAllTraitsByDefault = false;
        mixpanel.options.superProperties = ['accountStatus', 'subscribed', 'email'];
        analytics.identify(123, { accountStatus: 'Paid', subscribed: true, email: 'androidjones@sky.net' });
        analytics.called(window.mixpanel.identify, 123);
        analytics.called(window.mixpanel.register, { accountStatus: 'Paid', subscribed: true, $email: 'androidjones@sky.net' });
      });

      it('should set people properties from the mixpanel.options.peopleProperties object if setAllTraitsByDefault is false', function() {
        mixpanel.options.people = true;
        mixpanel.options.setAllTraitsByDefault = false;
        mixpanel.options.peopleProperties = ['friend'];
        analytics.identify(123, { friend: 'elmo' });
        analytics.called(window.mixpanel.identify, 123);
        analytics.called(window.mixpanel.people.set, { friend: 'elmo' });
      });

      it('should set people properties from the Mixpanel\'s special traits if setAllTraitsByDefault is false and the property isn\'t on the call', function() {
        mixpanel.options.people = true;
        mixpanel.options.setAllTraitsByDefault = false;
        mixpanel.options.peopleProperties = ['friend'];
        analytics.identify(123, { friend: 'elmo', email: 'dog@dog.com' });
        analytics.called(window.mixpanel.identify, 123);
        analytics.called(window.mixpanel.people.set, { friend: 'elmo', $email: 'dog@dog.com' });
      });


      it('shouldn\'t set super properties that aren\'t included in the mixpanel.options.superProperties object when setAllTraitsByDefault is false', function() {
        mixpanel.options.setAllTraitsByDefault = false;
        mixpanel.options.superProperties = ['accountStatus'];
        analytics.identify(123, { accountStatus: 'Paid', subscribed: true });
        analytics.called(window.mixpanel.identify, 123);
        analytics.called(window.mixpanel.register, { accountStatus: 'Paid' });
        analytics.didNotCall(window.mixpanel.register, { subscribed: true });
      });

      it('shouldn\'t set people properties from the mixpanel.options.peopleProperties object if setAllTraitsByDefault is false', function() {
        mixpanel.options.people = false;
        mixpanel.options.setAllTraitsByDefault = false;
        analytics.identify(123, { friend: 'elmo' });
        analytics.called(window.mixpanel.identify, 123);
        analytics.didNotCall(window.mixpanel.register);
        analytics.didNotCall(window.mixpanel.people.set);
      });

      it('should alias traits', function() {
        var date = new Date();
        analytics.identify({
          created: date,
          email: 'name@example.com',
          firstName: 'first',
          lastName: 'last',
          lastSeen: date,
          name: 'name',
          username: 'username',
          phone: 'phone'
        });
        analytics.called(window.mixpanel.register, {
          $created: iso(date),
          $email: 'name@example.com',
          $first_name: 'first',
          $last_name: 'last',
          $last_seen: iso(date),
          $name: 'name',
          $username: 'username',
          $phone: 'phone'
        });
      });

      it('should alias traits to Mixpanel People', function() {
        mixpanel.options.people = true;
        var date = new Date();
        analytics.identify({
          created: date,
          email: 'name@example.com',
          firstName: 'first',
          lastName: 'last',
          lastSeen: date,
          name: 'name',
          username: 'username',
          phone: 'phone'
        });
        analytics.called(window.mixpanel.people.set, {
          $created: iso(date),
          $email: 'name@example.com',
          $first_name: 'first',
          $last_name: 'last',
          $last_seen: iso(date),
          $name: 'name',
          $username: 'username',
          $phone: 'phone'
        });
      });

      it('should remove .created_at', function() {
        mixpanel.options.people = true;
        var date = new Date();
        analytics.identify({
          created_at: date,
          email: 'name@example.com',
          firstName: 'first',
          lastName: 'last',
          lastSeen: date,
          name: 'name',
          username: 'username',
          phone: 'phone'
        });
        analytics.called(window.mixpanel.people.set, {
          $created: iso(date),
          $email: 'name@example.com',
          $first_name: 'first',
          $last_name: 'last',
          $last_seen: iso(date),
          $name: 'name',
          $username: 'username',
          $phone: 'phone'
        });
      });

      it('should union arrays', function() {
        mixpanel.options.people = true;
        analytics.identify({
          pages_visited: ['homepage']
        });
        analytics.called(window.mixpanel.people.union, {
          pages_visited: ['homepage']
        });
      });

      it('should union arrays when setAllTraitsByDefault is false', function() {
        mixpanel.options.setAllTraitsByDefault = false;
        mixpanel.options.people = true;
        mixpanel.options.peopleProperties = ['pages_visited'];  // eslint-disable-line
        analytics.identify({
          pages_visited: ['homepage']
        });
        analytics.called(window.mixpanel.people.union, {
          pages_visited: ['homepage']
        });
      });
    });

    describe('#track', function() {
      beforeEach(function() {
        analytics.stub(window.mixpanel, 'track');
        analytics.stub(window.mixpanel, 'register');
        analytics.stub(window.mixpanel.people, 'increment');
        analytics.stub(window.mixpanel.people, 'set');
        analytics.stub(window.mixpanel.people, 'track_charge');
      });

      it('should send an event', function() {
        analytics.track('event');
        analytics.called(window.mixpanel.track, 'event');
      });

      it('should send an event and properties', function() {
        analytics.track('event', { property: true });
        analytics.called(window.mixpanel.track, 'event', { property: true });
      });

      it('should send a revenue property to Mixpanel People', function() {
        mixpanel.options.people = true;
        analytics.track('event', { revenue: 9.99 });
        analytics.called(window.mixpanel.people.track_charge, 9.99);
      });

      it('should set super properties from the mixpanel.options.superProperties object', function() {
        mixpanel.options.superProperties = ['accountStatus', 'subscribed', 'email'];
        analytics.track('event', { accountStatus: 'Paid', subscribed: true, email: 'androidjones@sky.net' });
        analytics.called(window.mixpanel.register, { accountStatus: 'Paid', subscribed: true, email: 'androidjones@sky.net' });
      });

      // Exercise a bug where we set people properties in track calls. https://segment.phacility.com/T903
      it('should not set people properties from the mixpanl.options.peopleProperties object', function() {
        mixpanel.options.people = true;
        mixpanel.options.peopleProperties = ['friend'];
        analytics.track('event', { friend: 'elmo' });
        analytics.didNotCall(window.mixpanel.people.set);
      });

      // Exercise a bug where we set people properties in track calls. https://segment.phacility.com/T903
      it('should not set people properties from the Mixpanel\'s special traits', function() {
        mixpanel.options.people = true;
        mixpanel.options.peopleProperties = ['friend'];
        analytics.track('event', { friend: 'elmo', email: 'dog@dog.com' });
        analytics.didNotCall(window.mixpanel.people.set);
      });

      // Exercise a bug where we set people properties in track calls. https://segment.phacility.com/T903
      it('shouldn\'t try to register super properties if not specified', function() {
        mixpanel.options.superProperties = [];
        analytics.track('event', { friend: 'elmo' });
        analytics.didNotCall(window.mixpanel.register);
      });

      it('should convert dates to iso strings', function() {
        var date = new Date();
        analytics.track('event', { date: date });
        analytics.called(window.mixpanel.track, 'event', { date: iso(date) });
      });

      it('should convert dates inside lists to iso strings', function() {
        var date = new Date();
        var date2 = new Date();
        analytics.track('event', { dates: [date, date2] });
        analytics.called(window.mixpanel.track, 'event', { dates: [iso(date), iso(date2)] });
      });

      it('should increment events that are in .increments option', function() {
        mixpanel.options.eventIncrements = [0, 'my event', 1];
        mixpanel.options.people = true;
        analytics.track('my event');
        analytics.called(window.mixpanel.people.increment, 'my event');
      });

      it('should should update people property if the event is in .increments', function() {
        mixpanel.options.eventIncrements = ['event'];
        mixpanel.options.people = true;
        analytics.track('event');
        var date = window.mixpanel.people.set.args[0][1];
        analytics.assert(date.getTime() === new Date().getTime());
        analytics.called(window.mixpanel.people.increment, 'event');
        analytics.called(window.mixpanel.people.set, 'Last event', date);
      });

      it('should not convert arrays of simple types', function() {
        analytics.track('event', { array: ['a', 'b', 'c'] });
        analytics.called(window.mixpanel.track, 'event', { array: ['a', 'b', 'c'] });
        analytics.track('event', { array: [13, 28, 99] });
        analytics.called(window.mixpanel.track, 'event', { array: [13, 28, 99] });
      });

      it('should remove mixpanel\'s reserved properties', function() {
        analytics.track('event', {
          distinct_id: 'string',
          ip: 'string',
          mp_name_tag: 'string',
          mp_note: 'string',
          token: 'string'
        });
        analytics.called(window.mixpanel.track, 'event', {});
      });

      it('should invert object list properties', function() {
        analytics.track('event', {
          products: [
            { price: 32, sku: 25 },
            { price: 33, sku: 29 }
          ]
        });
        analytics.called(window.mixpanel.track, 'event', {
          products_skus: [25, 29],
          products_prices: [32, 33]
        });
      });

      it('should increment properties when specified', function() {
        mixpanel.options.propIncrements = ['videos_watched'];
        mixpanel.options.people = true;
        analytics.track('event', {
          videos_watched: 3 
        });
        analytics.called(window.mixpanel.people.increment, 'videos_watched', 3);
      });
    });

    describe('#alias', function() {
      beforeEach(function() {
        analytics.stub(window.mixpanel, 'alias');
      });

      it('should send a new id', function() {
        analytics.alias('new');
        analytics.called(window.mixpanel.alias, 'new');
      });

      it('should send a new and old id', function() {
        analytics.alias('new', 'old');
        analytics.called(window.mixpanel.alias, 'new', 'old');
      });
    });
  });
});


lib/index.js
'use strict';

/**
 * Module dependencies.
 */

var alias = require('@segment/alias');
var dates = require('@segment/convert-dates');
var del = require('obj-case').del;
var includes = require('@ndhoule/includes');
var integration = require('@segment/analytics.js-integration');
var iso = require('@segment/to-iso-string');
var pick = require('@ndhoule/pick');
var is = require('is');
var indexOf = require('component-indexof');

/**
 * Expose `Mixpanel` integration.
 */

var Mixpanel = module.exports = integration('Mixpanel')
  .global('mixpanel')
  .option('eventIncrements', [])
  .option('propIncrements', [])
  .option('peopleProperties', [])
  .option('superProperties', [])
  .option('cookieName', '')
  .option('crossSubdomainCookie', false)
  .option('secureCookie', false)
  .option('persistence', 'cookie')
  .option('nameTag', true)
  .option('pageview', false)
  .option('people', false)
  .option('token', '')
  .option('setAllTraitsByDefault', true)
  .option('consolidatedPageCalls', true)
  .option('trackAllPages', false)
  .option('trackNamedPages', false)
  .option('trackCategorizedPages', false)
  .tag('<script src="//cdn.mxpnl.com/libs/mixpanel-2-latest.min.js">');

/**
 * Options aliases.
 */

var optionsAliases = {
  cookieName: 'cookie_name',
  crossSubdomainCookie: 'cross_subdomain_cookie',
  secureCookie: 'secure_cookie'
};

/**
 * Initialize.
 *
 * https://mixpanel.com/help/reference/javascript#installing
 * https://mixpanel.com/help/reference/javascript-full-api-reference#mixpanel.init
 *
 * @api public
 */

Mixpanel.prototype.initialize = function() {
  /* eslint-disable */
  (function(e,a){if(!a.__SV){var b=window;try{var c,l,i,j=b.location,g=j.hash;c=function(a,b){return(l=a.match(RegExp(b+"=([^&]*)")))?l[1]:null};g&&c(g,"state")&&(i=JSON.parse(decodeURIComponent(c(g,"state"))),"mpeditor"===i.action&&(b.sessionStorage.setItem("_mpcehash",g),history.replaceState(i.desiredHash||"",e.title,j.pathname+j.search)))}catch(m){}var k,h;window.mixpanel=a;a._i=[];a.init=function(b,c,f){function e(b,a){var c=a.split(".");2==c.length&&(b=b[c[0]],a=c[1]);b[a]=function(){b.push([a].concat(Array.prototype.slice.call(arguments,
0)))}}var d=a;"undefined"!==typeof f?d=a[f]=[]:f="mixpanel";d.people=d.people||[];d.toString=function(b){var a="mixpanel";"mixpanel"!==f&&(a+="."+f);b||(a+=" (stub)");return a};d.people.toString=function(){return d.toString(1)+".people (stub)"};k="disable time_event track track_pageview track_links track_forms register register_once alias unregister identify name_tag set_config reset people.set people.set_once people.increment people.append people.union people.track_charge people.clear_charges people.delete_user".split(" ");
for(h=0;h<k.length;h++)e(d,k[h]);a._i.push([b,c,f])};a.__SV=1.2;}})(document,window.mixpanel||[]);
  /* eslint-enable */
  this.options.eventIncrements = lowercase(this.options.eventIncrements);
  this.options.propIncrements = lowercase(this.options.propIncrements);
  var options = alias(this.options, optionsAliases);
  // tag ajs requests with Segment by request from Mixpanel team for better mutual debugging
  options.loaded = function(mixpanel) {
    mixpanel.register({ mp_lib: 'Segment: web' });
  };
  window.mixpanel.init(options.token, options);
  this.load(this.ready);
};

/**
 * Loaded?
 *
 * @api private
 * @return {boolean}
 */

Mixpanel.prototype.loaded = function() {
  return !!(window.mixpanel && window.mixpanel.config);
};

/**
 * Page.
 *
 * https://mixpanel.com/help/reference/javascript-full-api-reference#mixpanel.track_pageview
 *
 * @api public
 * @param {Page} page
 */

Mixpanel.prototype.page = function(page) {
  var category = page.category();
  var name = page.name();
  var opts = this.options;

  // consolidated Page Calls
  if (opts.consolidatedPageCalls) {
    this.track(page.track());
    return;
  }

  // all pages
  if (opts.trackAllPages) {
    this.track(page.track());
    return;
  }

  // categorized pages
  if (opts.trackCategorizedPages && category) {
    // If this option is checked and name was also passed, used the full name which includes both category & name
    if (name) {
      this.track(page.track(page.fullName()));
      return;
    }
    this.track(page.track(category));
    return;
  }

  // named pages
  if (name && opts.trackNamedPages) {
    this.track(page.track(name));
  }
};

/**
 * Trait aliases.
 */

var traitAliases = {
  created: '$created',
  email: '$email',
  firstName: '$first_name',
  lastName: '$last_name',
  lastSeen: '$last_seen',
  name: '$name',
  username: '$username',
  phone: '$phone'
};

/**
 * Identify.
 *
 * https://mixpanel.com/help/reference/javascript#super-properties
 * https://mixpanel.com/help/reference/javascript#user-identity
 * https://mixpanel.com/help/reference/javascript#storing-user-profiles
 *
 * @api public
 * @param {Identify} identify
 */

Mixpanel.prototype.identify = function(identify) {
  var username = identify.username();
  var email = identify.email();
  var id = identify.userId();
  var setAllTraitsByDefault = this.options.setAllTraitsByDefault;
  var people = this.options.people;
  var peopleProperties = extendTraits(this.options.peopleProperties);
  var superProperties = this.options.superProperties;

  // id
  if (id) window.mixpanel.identify(id);

  // name tag
  var nametag = email || username || id;
  if (nametag) window.mixpanel.name_tag(nametag);

  var traits = identify.traits(traitAliases);
  if (traits.$created) del(traits, 'createdAt');
  traits = dates(traits, iso);

  // determine which traits to union to existing properties and which to set as new properties
  var traitsToUnion = {};
  var traitsToSet = {};
  for (var key in traits) {
    if (!traits.hasOwnProperty(key)) continue;

    var trait = traits[key];
    if (Array.isArray(trait) && trait.length > 0) {
      traitsToUnion[key] = trait;
      // since mixpanel doesn't offer a union method for super properties we have to do it manually by retrieving the existing list super property 
      // from mixpanel and manually unioning to it ourselves
      var existingTrait = window.mixpanel.get_property(key);
      if (existingTrait && Array.isArray(existingTrait)) {
        traits[key] = unionArrays(existingTrait, trait);
      }
    } else {
      traitsToSet[key] = trait;
    }
  }

  if (setAllTraitsByDefault) {
    window.mixpanel.register(traits);
    if (people) {
      window.mixpanel.people.set(traitsToSet);
      window.mixpanel.people.union(traitsToUnion);
    }
  } else {
    // explicitly set select traits as people and super properties
    var mappedSuperProps = mapTraits(superProperties);
    var superProps = pick(mappedSuperProps || [], traits);
    if (!is.empty(superProps)) window.mixpanel.register(superProps);
    if (people) {
      var mappedPeopleProps = mapTraits(peopleProperties);
      var peoplePropsToSet = pick(mappedPeopleProps || [], traitsToSet);
      var peoplePropsToUnion = pick(mappedPeopleProps || [], traitsToUnion);
      if (!is.empty(peoplePropsToSet)) window.mixpanel.people.set(peoplePropsToSet);
      if (!is.empty(peoplePropsToUnion)) window.mixpanel.people.union(peoplePropsToUnion);
    }
  }
};

/**
 * Track.
 *
 * https://mixpanel.com/help/reference/javascript#sending-events
 * https://mixpanel.com/help/reference/javascript#tracking-revenue
 *
 * @api public
 * @param {Track} track
 */

Mixpanel.prototype.track = function(track) {
  var eventIncrements = this.options.eventIncrements || this.options.increments; // TODO: remove settings.increments check, it's only here as we cutover from increments to eventIncrements
  var propIncrements = this.options.propIncrements;
  var eventLowercase = track.event().toLowerCase();
  var people = this.options.people;
  var props = track.properties();
  var revenue = track.revenue();
  // Don't map traits, clients should use identify instead.
  var superProps = pick(this.options.superProperties, props);

  // delete mixpanel's reserved properties, so they don't conflict
  delete props.distinct_id;
  delete props.ip;
  delete props.mp_name_tag;
  delete props.mp_note;
  delete props.token;

  props = dates(props, iso);
  invertObjectArrays(props);

  // Mixpanel People operations
  if (people) {
    // increment event count
    if (includes(eventLowercase, eventIncrements)) {
      window.mixpanel.people.increment(track.event());
      window.mixpanel.people.set('Last ' + track.event(), new Date());
    }
    // increment property counts
    for (var key in props) {
      if (!Object.prototype.hasOwnProperty.call(props, key)) {
        continue;
      }
      var prop = props[key];
      if (includes(key.toLowerCase(), propIncrements)) {
        window.mixpanel.people.increment(key, prop);
      }
    }
    // track revenue
    if (revenue) {
      window.mixpanel.people.track_charge(revenue);
    }
  }

  // track the event
  var query;
  if (props.link_query) {
    query = props.link_query; // DOM query
    delete props.link_query;
    window.mixpanel.track_links(query, track.event(), props);
  } else if (props.form_query) {  // DOM query
    query = props.form_query;
    delete props.form_query;
    window.mixpanel.track_forms(query, track.event(), props);
  } else {
    window.mixpanel.track(track.event(), props);
  }

  // register super properties if present in context.mixpanel.superProperties
  if (!is.empty(superProps)) {
    window.mixpanel.register(superProps);
  }
};

/**
 * Alias.
 *
 * https://mixpanel.com/help/reference/javascript#user-identity
 * https://mixpanel.com/help/reference/javascript-full-api-reference#mixpanel.alias
 *
 * @api public
 * @param {Alias} alias
 */

Mixpanel.prototype.alias = function(alias) {
  var mp = window.mixpanel;
  var to = alias.to();
  if (mp.get_distinct_id && mp.get_distinct_id() === to) return;
  // HACK: internal mixpanel API to ensure we don't overwrite
  if (mp.get_property && mp.get_property('$people_distinct_id') === to) return;
  // although undocumented, mixpanel takes an optional original id
  mp.alias(to, alias.from());
};

/**
 * Lowercase the given `arr`.
 *
 * @api private
 * @param {Array} arr
 * @return {Array}
 */

function lowercase(arr) {
  var ret = new Array(arr.length);

  for (var i = 0; i < arr.length; ++i) {
    ret[i] = String(arr[i]).toLowerCase();
  }

  return ret;
}

/**
 * Map Special traits in the given `arr`.
 * From the TraitAliases for Mixpanel's special props
 *
 * @api private
 * @param {Array} arr
 * @return {Array}
 */

function mapTraits(arr) {
  var ret = new Array(arr.length);

  for (var i = 0; i < arr.length; ++i) {
    if (traitAliases.hasOwnProperty(arr[i])) {
      ret.push(traitAliases[arr[i]]);
    } else {
      ret.push(arr[i]);
    }
  }

  return ret;
}

/**
 * extend Mixpanel's special trait keys in the given `arr`.
 *
 * @api private
 * @param {Array} arr
 * @return {Array}
 */

function extendTraits(arr) {
  var keys = [];

  for (var key in traitAliases) {
    if (traitAliases.hasOwnProperty(key)) {
      keys.push(key);
    }
  }

  for (var i = 0; i < keys.length; ++i) {
    if (indexOf(arr, keys[i]) < 0) {
      arr.push(keys[i]);
    }
  }

  return arr;
}

/**
 * Since Mixpanel doesn't support lists of objects, invert each list of objects to a set of lists of object properties.
 * Treats list transformation atomically, e.g. will only transform if EVERY item in list is an object
 *
 * @api private
 * @param {Object} props
 * @example 
 * input: {products: [{sku: 32, revenue: 99}, {sku:2, revenue: 103}]}
 * output: {products_skus: [32, 2], products_revenues: [99, 103]}
 */

function invertObjectArrays(props) {
  for (var propName in props) {  // eslint-disable-line
    var propValue = props[propName];
    if (!props.hasOwnProperty(propName) || !Array.isArray(propValue)) {
      continue;
    }

    var invertedArrays = invertObjectArray(propName, propValue);
    if (Object.keys(invertedArrays).length !== 0) { // make sure obj isn't empty
      mergeArraysIntoObj(props, invertedArrays);
      delete props[propName];
    }
  }
}

// Example:
// input: 'products', [{sku: 32, revenue: 99}, {sku:2, revenue: 103}]
// output: {products_skus: [32, 2], products_revenues: [99, 103]}
function invertObjectArray(propName, arr) {
  var invertedArrays = {};

  // invert object lists and collect into invertedLists
  for (var i=0; i<arr.length; i++) {
    var elem = arr[i];

    // abort operation if non-object encountered in array
    if (typeof elem !== 'object') {
      return {};
    }
    for (var key in elem) {
      if (!elem.hasOwnProperty(key)) {
        continue;
      }
      var attrKey = propName+'_'+key+'s';  // e.g. products_skus
      
      // append to list if it exists or create new one if not
      if (attrKey in invertedArrays) {
        invertedArrays[attrKey].push(elem[key]);
      } else {
        invertedArrays[attrKey] = [elem[key]];
      }
    }
  }
  return invertedArrays;
}

function mergeArraysIntoObj(destination, source) {
  for (var arrayName in source) {
    if (source.hasOwnProperty(arrayName)) {
      var arr = source[arrayName];
      destination[arrayName] = arrayName in destination ? destination[arrayName].concat(arr) : arr;
    }
  }
}


/**
 * Return union of two arrays
 * Pulled from https://stackoverflow.com/a/3629861
 *
 * @param {Array} x
 * @param {Array} y
 * @return {Array} res
 * @api private
 */

function unionArrays(x, y) {
  var obj = {};
  // store items of each array as keys/values of obj, implicitly overwriting duplicates
  var i;
  for (i = 0; i < x.length; i++) {
    obj[x[i]] = x[i];
  }
  for (i = 0; i < y.length; i++) {
    obj[y[i]] = y[i];
  }
  return Object.keys(obj);
}